ENH: add mean keyword to std and var

RonaldAJ opened the pull request:
Replaces previous pull request  #23889 (ENH: Introduce new functions mean_std and mean_var.)  after triage review.

Often when the standard deviation is needed the mean is also needed; the same holds for the variance and the mean. With the current code the mean is then calculated twice, this can be prevented if the functions calculating the variance or the standard deviation can use a precalculated mean. See ticket https://github.com/numpy/numpy/issues/23741. 

These changes can be largely independent from those for ticket https://github.com/numpy/numpy/issues/13199, which suggests reducing memory usage by avoiding storing data-mean in a separate intermediate array.

See also:  [discussion on the mailing list](https://mail.python.org/archives/list/numpy-discussion@python.org/thread/TJI6AKD2NT7XSBJTGGJXAQPPLBKGX3XN/)

RonaldAJ committed:
Add mean keyword to std and var functions.

RonaldAJ committed:
Add releae note for mean keyword to std and var functions.

RonaldAJ committed:
Update release note with PR number

RonaldAJ committed:
Address lint issue.

RonaldAJ committed:
Align nan signatures with new signatures.

RonaldAJ committed:
Address lint issue.

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -3506,13 +3506,13 @@ def mean(a, axis=None, dtype=None, out=None, keepdims=np._NoValue, *,
 def _std_dispatcher(a, axis=None, dtype=None, out=None, ddof=None,
-                    keepdims=None, *, where=None):
+                    keepdims=None, *, where=None, mean=None):
     return (a, where, out)
```
In principle, one should dispatch on `mean` if given, so `return (a, where, out, mean)`

RonaldAJ replied:
To be honest it is not clear to me why these dispatchers are needed and I merely followed the pattern. 

By making mean a keyword input it is no longer intended to be an ouput aswell. So why should it be returned?

mhvk replied:
The dispatcher returns all arguments that can, in principle, contain something that is an array and hence that, if not a regular `ndarray`, can allow the function to be dealt with another package (say, `dask`). Since `mean` can be an array, it should be added (most similar to `where`).

RonaldAJ replied:
I see, I'll make the change then.

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -3647,13 +3654,13 @@ def std(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
 def _var_dispatcher(a, axis=None, dtype=None, out=None, ddof=None,
-                    keepdims=None, *, where=None):
+                    keepdims=None, *, where=None, mean=None):
     return (a, where, out)
```
Same here, `return (a, where, out, mean)`

RonaldAJ replied:
I'll make the change.

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -3558,6 +3558,10 @@ def std(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
+          and the axis the same as used in the call to this std function. 
+          For masked arrays this argument is ignored.
 
         .. versionadded:: 1.20.0
```
This `versionadded` belongs to `where` - you should add a new one with `1.26.0`.

RonaldAJ replied:
Corrected locally now.

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -3700,7 +3707,11 @@ def var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
+          and the axis the same as used in the call to this var function.
+          For masked arrays this argument is ignored. 
+          
         .. versionadded:: 1.20.0
```
As for `std`

RonaldAJ replied:
Corrected locally now.

mhvk reviewed (commented):
This seems a good solution! I also like that now one can put in something else than the mean from the same data (which is often useful). But see the important comment about subclasses, plus a lot of more nitpicky ones.

Note that I'm surprised `MaskedArray` does not work. Did you try again? Could you explain the problem? And is this a new problem, or did the functions not work even before? If they worked before, it would be much nicer to get `MaskedArray` right too.

RonaldAJ committed:
Correct version numbers on keywords.

RonaldAJ commented on doc/release/upcoming_changes/24126.change.rst:
```diff
@@ -0,0 +1,27 @@
+                 keepdims=False,
+                 mean=mean)
+
+Note that the mean passed into the std or var function needs to be calculated with keepdims=True. 
```
Corrected locally now for both std and var.

RonaldAJ committed:
Put backticks on keyword argument in documentation string.

RonaldAJ commented on numpy/core/fromnumeric.py:
```diff
@@ -3774,6 +3785,8 @@ def var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
         kwargs['keepdims'] = keepdims
     if where is not np._NoValue:
         kwargs['where'] = where
+    if mean is not np._NoValue and type(a) is mu.ndarray:
```
I guess that what you suggest is somewhat similar to what was suggested during the triage review regarding the shape of the mean. Don't handle it, let other mechanisms raise errors if it is used outside of the intended scope.

So I would just drop `and type(a) is mu.ndarray` here?

RonaldAJ replied:
I dropped the second part of the test now.

RonaldAJ commented on numpy/core/tests/test_numeric.py:
```diff
@@ -298,6 +299,320 @@ def test_var(self):
+        # Output should be the same as from the individual algorithms
+        std_old = np.std(A, axis=1, keepdims=True)
+
+        assert_equal(np.array(std_old.shape), np.array(mean.shape))
```
This specific case compares the shapes of the mean and the std. Those should be the same if calculated with the  same keepdims value. But he arrays should be different.

RonaldAJ replied:
Where followed by comparison of the full matrix content I now dropped the shape comparison.

RonaldAJ committed:
Cleanuup assert statements in tests

RonaldAJ commented on numpy/core/tests/test_numeric.py:
```diff
@@ -298,6 +299,320 @@ def test_var(self):
+                     mean=mean)
+
+        # Shape of returned mean and std should be same
+        assert_equal(np.array(std.shape), np.array(mean.shape))
```
I guess that would make reading easier. Corrected locally now.

RonaldAJ commented on numpy/core/tests/test_numeric.py:
```diff
@@ -298,6 +299,320 @@ def test_var(self):
+        assert_equal(np.array(std_old.shape), np.array(mean.shape))
+        assert_almost_equal(std, std_old)
+
+        # The returned  objects should be the objects specified during calling
```
Corrected locally now.

RonaldAJ committed:
Move comparison of in and out arrays closer to the function call.

RonaldAJ committed:
Remove clutter from example code in release note.

RonaldAJ commented on doc/release/upcoming_changes/24126.change.rst:
```diff
@@ -0,0 +1,27 @@
+                   keepdims=True)
+
+    std = np.std(A,
+                 out=None,
```
Corrected locally now.

RonaldAJ committed:
Add test for nanstd and fix error in nanvar

RonaldAJ committed:
haqndle "mean" keyword for var and std on MaskedArrays.

RonaldAJ commented:
> Note that I'm surprised `MaskedArray` does not work. Did you try again? Could you explain the problem? And is this a new problem, or did the functions not work even before? If they worked before, it would be much nicer to get `MaskedArray` right too.

I now added support for "mean" keyword on MaskedArrays.

RonaldAJ committed:
Address lint issues.

RonaldAJ committed:
update the dispatchers according to suggestions by Marten van Kerkwijk:

(https://github.com/numpy/numpy/pull/24126#discussion_r1254314302) The dispatcher returns all arguments that can, in principle, contain something that is an array and hence that, if not a regular ndarray, can allow the function to be dealt with another package (say, dask). Since mean can be an array, it should be added (most similar to where).

RonaldAJ commented:
@charris The cygwin test failed, but from the reported error I guess it has little to do with my change. Will it run again automagically?

mattip commented:
I re-ran it manually. If it fails again we can ignore the failure.

mattip commented on doc/release/upcoming_changes/24126.change.rst:
```diff
@@ -0,0 +1,15 @@
+Add mean keyword to var and std function
+----------------------------------------
+
+Often when the standard deviation is needed the mean is also needed; the same holds for the variance and the mean. With the current code the mean is then calculated twice, the change introduced here for the var and std functions allows for passing in a precalculated mean as an keyword argument.
```
Can you reformat to ~80 columns? Also: no need to make the release note extensive, it can reference the docstring and just be a few lines. The examples should move to the docstring and be more complete, like the other examples.

RonaldAJ replied:
Moved and adjusted the example to the doc-string and reflowed the release note.

mattip commented on numpy/core/fromnumeric.py:
```diff
@@ -3560,6 +3560,12 @@ def std(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
 
         .. versionadded:: 1.20.0
+        
+    mean : optionally provide the mean to prevent recalculation, the mean
```
Nit: I don't like long sentences. If you decide to go with my suggestion, please do it everywhere. Also, some of the docstrings do not have quotes around `keepdims = True`
```suggestion
    mean : optionally provide the mean to prevent recalculation. The mean
```

RonaldAJ replied:
The quotes are there now, also added description of the type and the label optional.

mattip commented on numpy/core/fromnumeric.py:
```diff
@@ -3700,8 +3709,15 @@ def var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
         Elements to include in the variance. See `~numpy.ufunc.reduce` for
         details.
-
+        
```
```suggestion

```

RonaldAJ replied:
I hope I fixed it.

RonaldAJ replied:
.. versionadded:: 1.20.0

Appears with two white lines around it. I copied that.

mattip commented on numpy/core/fromnumeric.py:
```diff
@@ -3560,6 +3560,12 @@ def std(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue, *,
         See `~numpy.ufunc.reduce` for details.
 
         .. versionadded:: 1.20.0
+        
```
Remove trailing spaces from blank lines
```suggestion

```

RonaldAJ replied:
Done.

mattip reviewed (commented):
Looks good, @mhvk already did all the review. Just a few nits about the release note and docstrings.

RonaldAJ committed:
Move and adjust example from release note to doc-strings. Reflow doc-string.

RonaldAJ committed:
Improve doc-string. Shorter sentences and add type and label mean argument

RonaldAJ committed:
Remove some of these pesky trailing white spaces

RonaldAJ committed:
Make extra white lines more consistent.

mattip commented on numpy/lib/nanfunctions.py:
```diff
@@ -1824,6 +1834,10 @@ def nanstd(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue,
+    mean : optionally provide the mean to prevent recalculation, the mean
+          should have a shape as if it was calculated with keepdims = True
+          and the axis the same as used in the call to this var function.
+          For masked arrays this argument is ignored.
```
This needs a bit of tweaking like the other ones. As well, the 1.22.0 version added line belongs to `where`.

mattip commented on numpy/lib/nanfunctions.py:
```diff
@@ -1824,6 +1834,10 @@ def nanstd(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue,
+          and the axis the same as used in the call to this var function.
+          For masked arrays this argument is ignored.
 
         .. versionadded:: 1.22.0
```
This needs a bit of tweaking like the other ones. As well, the 1.22.0 version added line belongs to `where`.

RonaldAJ committed:
Make sure code examples execute without Jupyter magic.

RonaldAJ committed:
Fold lines to pass linter.

RonaldAJ committed:
Update doc-string nanstd and nanvar.

RonaldAJ committed:
Try to satisfy linter and apple requirements at the same time. Making the example code ugly, alas!

RonaldAJ commented:
The output of the examples depends on platform and circumstances. There fore testing based on example code gives errors:

`--------------
2023-07-06T19:09:55.1262830Z 
2023-07-06T19:09:55.1263570Z File "build/testenv/lib/python3.9/site-packages/numpy/__init__.py", line ?, in std
2023-07-06T19:09:55.1264100Z Failed example:
2023-07-06T19:09:55.1264890Z     print(f'Percentage execution time saved {100*(t2-t1)/t2:.0f}%')
2023-07-06T19:09:55.1265550Z Expected:
2023-07-06T19:09:55.1266030Z     Percentage execution time saved 30%
2023-07-06T19:09:55.1266470Z Got:
2023-07-06T19:09:55.1266880Z     Percentage execution time saved 32%
2023-07-06T19:09:55.1267080Z 
2023-07-06T19:09:55.1267310Z numpy.core.var
2023-07-06T19:09:55.1267770Z --------------
2023-07-06T19:09:55.1267930Z 
2023-07-06T19:09:55.1268440Z File "build/testenv/lib/python3.9/site-packages/numpy/__init__.py", line ?, in var
2023-07-06T19:09:55.1268800Z Failed example:
2023-07-06T19:09:55.1269300Z     print(f'Percentage execution time saved {100*(t2-t1)/t2:.0f}%')
2023-07-06T19:09:55.1269630Z Expected:
2023-07-06T19:09:55.1275670Z     Percentage execution time saved 32%
2023-07-06T19:09:55.1276120Z Got:
2023-07-06T19:09:55.1276900Z     Percentage execution time saved 31%`

Is there a method for handling this?

RonaldAJ committed:
Make doctest skip resource dependent output

RonaldAJ commented:
Added:  #doctest: +SKIP

mattip commented:
> Is there a method for handling this?

You could do something like
```
# Using mean saves around 32% of the compute time
>>> time_saved = (t2-t1)/t2
>>> assert 0.3 < time_saved < 0.35
```

RonaldAJ commented:
> > Is there a method for handling this?
> 
> You could do something like
> 
> ```
> # Using mean saves around 32% of the compute time
> >>> time_saved = (t2-t1)/t2
> >>> assert 0.3 < time_saved < 0.35
> ```

 #doctest: +SKIP did the trick

mattip commented:
Thanks @RonaldAJ and @mhvk. This is a nice API addition.

RonaldAJ commented:
I think this issue https://github.com/numpy/numpy/issues/23741 can be closed.
